src/policy/immix/immixspace.rs

```diff
@@ -618,11 +620,11 @@ impl<VM: VMBinding> ImmixSpace<VM> {
         } else if self.is_marked(object) {
             // We won the forwarding race but the object is already marked so we clear the
             // forwarding status and return the unmoved object
-            debug_assert!(
-                nursery_collection || self.defrag.space_exhausted() || self.is_pinned(object),
-                "Forwarded object is the same as original object {} even though it should have been copied",
-                object,
-            );
+            // debug_assert!(
+            //     nursery_collection || self.defrag.space_exhausted() || self.is_pinned(object),
+            //     "Forwarded object is the same as original object {} even though it should have been copied",
+            //     object,
+            // );
```

Comment:
I'm not sure I understand how to do this. This would trigger also after then TP closure (in fact, only after), when the objects have been marked already by the TP closure, which is the only reason why we're not moving the object. When we're doing the TP closure, we call into `trace_object_without_moving`.

Reply:
You could add a parameter to the method (like the `nursery_collection` parameter) so we know whether this is in a TP closure and use that in the assertion. Probably discuss with Kunshan on this.

Reply:
But this will run _after_ the TP closure, as far as I understand. The TP closure will mark things in place, this code will see that these objects have been marked by the TP closure and therefore have not been moved - which would fail the assertion.

Reply:
IMO, the only way we could keep this assertion is when we're freely able to move objects opportunistically. However, as soon as we introduce pinning roots or transitively pinning roots, we cannot have the assertion, since the previous trace could have already marked the object in place (either the pinning or the transitively pinning one).

Reply:
I see.  Because the object may have been marked by a different worker, we no longer know why the object was pinned by that worker because that information was not recorded.  It was possible to do the assertion because the condition of when to move and when not to move was simple.  With TP closure introduced, it becomes a bit more dynamic.

I am OK with removing this assertion.